src/ring_theory/graded_algebra/homogeneous_localization.lean

```diff
@@ -0,0 +1,452 @@
+  a proof that `f.denom ∈ f.deg`.
+* `homogeneous_localization.eq_num_div_denom`: if `f : homogeneous_localization 𝒜 x`, then
+  `f.val : Aₓ` is equal to `f.num / f.denom`.
+
+## References
+
+* [Robin Hartshorne, *Algebraic Geometry*][Har77]
+
+
+-/
+
+noncomputable theory
+
+open_locale direct_sum big_operators pointwise
+open direct_sum set_like
+
+variables {ι R A: Type*}
+variables [add_comm_monoid ι] [decidable_eq ι]
+variables [comm_ring R] [comm_ring A] [algebra R A]
+variables (𝒜 : ι → submodule R A) [graded_algebra 𝒜]
+variables (x : ideal A) [ideal.is_prime x]
+
+local notation `at ` x := localization.at_prime x
+
+namespace homogeneous_localization
+
+section
+/--
+Let `x` be a prime ideal, then `num_denom_same_deg 𝒜 x` is a structure with a numerator and a
+denominator with same grading such that the denominator is not contained in `x`.
+-/
+@[nolint has_inhabited_instance]
+structure num_denom_same_deg :=
+(deg : ι)
+(num denom : 𝒜 deg)
+(denom_not_mem : (denom : A) ∉ x)
+
+end
+
+namespace num_denom_same_deg
+
+open set_like.graded_monoid submodule
+
+variable {𝒜}
+@[ext] lemma ext {c1 c2 : num_denom_same_deg 𝒜 x} (hdeg : c1.deg = c2.deg)
+  (hnum : (c1.num : A) = c2.num) (hdenom : (c1.denom : A) = c2.denom) :
+  c1 = c2 :=
+begin
+  rcases c1 with ⟨i1, ⟨n1, hn1⟩, ⟨d1, hd1⟩, h1⟩,
+  rcases c2 with ⟨i2, ⟨n2, hn2⟩, ⟨d2, hd2⟩, h2⟩,
+  dsimp only [subtype.coe_mk] at *,
+  simp only,
+  exact ⟨hdeg, by subst hdeg; subst hnum, by subst hdeg; subst hdenom⟩,
+end
+
+instance : has_one (num_denom_same_deg 𝒜 x) :=
+{ one :=
+  { deg := 0,
+    num := ⟨1, one_mem⟩,
+    denom := ⟨1, one_mem⟩,
+    denom_not_mem := λ r, (infer_instance : x.is_prime).ne_top $ x.eq_top_iff_one.mpr r } }
+
+@[simp] lemma deg_one : (1 : num_denom_same_deg 𝒜 x).deg = 0 := rfl
+@[simp] lemma num_one : ((1 : num_denom_same_deg 𝒜 x).num : A) = 1 := rfl
+@[simp] lemma denom_one : ((1 : num_denom_same_deg 𝒜 x).denom : A) = 1 := rfl
+
+instance : has_zero (num_denom_same_deg 𝒜 x) :=
+{ zero := ⟨0, 0, ⟨1, one_mem⟩, λ r, (infer_instance : x.is_prime).ne_top $ x.eq_top_iff_one.mpr r⟩ }
+
+@[simp] lemma deg_zero : (0 : num_denom_same_deg 𝒜 x).deg = 0 := rfl
+@[simp] lemma num_zero : (0 : num_denom_same_deg 𝒜 x).num = 0 := rfl
+@[simp] lemma denom_zero : ((0 : num_denom_same_deg 𝒜 x).denom : A) = 1 := rfl
+
+instance : has_mul (num_denom_same_deg 𝒜 x) :=
+{ mul := λ p q,
+  { deg := p.deg + q.deg,
+    num := ⟨p.num * q.num, mul_mem p.num.prop q.num.prop⟩,
+    denom := ⟨p.denom * q.denom, mul_mem p.denom.prop q.denom.prop⟩,
+    denom_not_mem := λ r, or.elim
+      ((infer_instance : x.is_prime).mem_or_mem r) p.denom_not_mem q.denom_not_mem } }
+
+lemma deg_mul (c1 c2 : num_denom_same_deg 𝒜 x) : (c1 * c2).deg = c1.deg + c2.deg := rfl
+lemma num_mul (c1 c2 : num_denom_same_deg 𝒜 x) : ((c1 * c2).num : A) = c1.num * c2.num := rfl
+lemma denom_mul (c1 c2 : num_denom_same_deg 𝒜 x) :
+  ((c1 * c2).denom : A) = c1.denom * c2.denom := rfl
```

Comment:
I think all the lemmas of this style may as well be `simp`.

Reply:
done